### DISCUSSION TITLE
An utility method to trigger a function when something changed to a ref, or nothing changed for x seconds

### DISCUSSION THREAD

**Discussion by bobmulder** [vueuse/vueuse#3355] [Ideas]
Hi all,

I have a usecase where I want to trigger a method when something is added to my ref, or the ref didn't change for x seconds. I asked ChatGPT for a suggestion:

https://chat.openai.com/share/b781d99c-26d2-4cc4-a8dd-f0d7ed88e555

```js
import { ref, watch, onUnmounted } from 'vue';

const myList = ref([]);
let timerId;

function triggerMethod() {
  console.log('Method triggered after 3 seconds of inactivity.');
  // Place your logic here
}

watch(myList, () => {
  clearTimeout(timerId);
  timerId = setTimeout(triggerMethod, 3000);
});

onUnmounted(() => {
  clearTimeout(timerId);
});
```

I have been looking through the vueuse list, and cant find any utility method which does something like this. Would it be useful to add this to vueuse, or not?

**Comment by darrylnoakes** [comment]
You (appear to) want "when changed, or no changes for x seconds." Effectively, a watch with a timeout.

That code watches for a change and calls the function 3 seconds later, clearing any previous timeout; effectively, it's a debounced watch: it will only call the function after a change if there have been no more changes in 3 seconds. That (with a max wait, as well) is available in the form of [`watchDebounced()`](https://vueuse.org/shared/watchdebounced/)

I assume you would want to keep triggering after successive timeouts. I.e., trigger every x seconds, or sooner if there is a change.
If not, you would need to state your exact required behavior.

Because of the way the callback is being called, I've made it so that a watch-triggered call has the standard arguments and a timeout-triggered call has the value of the source passed as the first two, and a no-op `onCleanup` function.
So here's my quick, naive POC:

```typescript
import {
  watch,
  type Ref,
  type WatchOptions,
} from "vue";

function watchWithPeriod<T>(
  ref: Ref<T>,
  cb: (
    value: T,
    oldValue: T | undefined,
    onCleanup: (cleanupFn: () => void) => void,
  ) => void,
  options: WatchOptions & { period: number },
) {
  let timer: number | undefined = setInterval(
    () => cb(ref.value, ref.value, () => {}),
    options.period,
  );

  watch(
    ref,
    (...args) => {
      clearTimeout(timer);
      timer = setInterval(
        () => cb(ref.value, ref.value, () => {}),
        options.period,
      );

      cb(...args);
    },
    options,
  );
}
```

It would be used like this:

```typescript
 import { ref } from "vue";

const msg = ref("");

watchWithPeriod(
  msg,
  (value, oldValue, onCleanup) => {
    if (value !== oldValue)
      console.log("Value changed:", `"${oldValue}"`, `"${value}"`);
    else
      console.log("Interval:", `"${value}"`)
    onCleanup(() => console.log("Cleanup called."));
  },
  { period: 3000 },
);
```

[Vue Playground](https://play.vuejs.org/#eNqdVE1v2zAM/SucMAwJkLkFevOSAFvRAd1hK7pgvehQx2YSd7IkyHLawfN/H/XlpEXWQ3OJyEdSj88Ue/ZZ62zfIcvZvC1NrS20aDsNopDbBWe25WzJZd1oZSz0XAIY3Mzc/2Nhy50/2T8a4Ta6vXHnsB/a1kq25B1gY1QDnNFNnH3ikstNJ0sHhzJ3td3doKlVNV8tJ/GW3NUk25ct1zl4AGBfiA5zWHk/gBLVr+iBv9DJCje1xCqh8lJgITtN6WU4fZV0nsJiCXtVV+PBJzy3VOggf9YPfIAetCebg+yaNRoYKHwa5BFowdYNmhE8IgULJ++1tGioi9hP4FKuJ9Rz5pubufbTMcD9ME0dBRpZoBBYe0njJ4lV02ei+lmWFWbbhjrBR4KSGGZFRFVnJ56wqxIwb57k+ga+/+Hsfom357MeeUYeTtSj5LHTwSWV5LHQtFuiSZdPOOMslHsxUJ44xfn0SeSYZoZOaT6O1ak3MRLeLRZj8HSkSncrgZlQW7rXY1Du6MFglXM2g3vO3vcpa+DsPrp8SWenDlG0eLpokv1Q75D8YrAn8Xs8y48YlIUQWGWkTbgzaHqY34vz8/MwvQTPz8IKoAdPhsVGi8IiWQBzvex7L/cwzM908NVSdxb2HxtVoaBlQbDfFvOzMZfNaIUQtU29zR5aJWnPeI05K1Wja4EmvirO8qQ+Z0RaPX7zPmvoKyV/ucPy9wn/Q/vkfJzdGGxJN1oyI2ZppNAG+Ornd3yi8wgS805Q9CvgLZKsneMYwr7QYybaR3Ge7bXfkLXcrtqrJ4uyTU05ogfl/Qq8fKX1A92L7MLn0cCz4R96Cs7C)

**Reply by bobmulder** [reply]
Thank you for your reply @darrylnoakes. You are right about "when changed, or no changes for x seconds.". For me, using [watchDebounced()](https://vueuse.org/shared/watchdebounced/) was not a feasable solution, cause I want to trigger my callback immediately when changed (any time). Maybe I did not understand the `watchDebounced()` method (eventhough I've been using it for a while). 

When would `onCleanup` be called in your POC? I don't understand its purpose 🙈

**Reply by darrylnoakes** [reply]
@bobmulder `onCleanup` is the [standard Vue `onCleanup` function](https://vuejs.org/api/reactivity-core.html#watch:~:text=The%20cleanup%20callback%20will%20be%20called%20right%20before%20the%20next%20time%20the%20effect%20is%20re%2Drun%2C%20and%20can%20be%20used%20to%20clean%20up%20invalidated%20side%20effects%2C%20e.g.%20a%20pending%20async%20request.), which is passed as the third argument to a watcher callback.

> The callback receives three arguments: the new value, the old value, and a function for registering a side effect cleanup callback. The cleanup callback will be called right before the next time the effect is re-run, and can be used to clean up invalidated side effects, e.g. a pending async request.

I included it to demonstrate that a function registered with it will only be called before a watch-triggered call, not a timer-triggered call.
